pkg/frontend/querymiddleware/results_cache.go

```diff
@@ -757,3 +738,111 @@ func mergeSamplesProcessedPerStep(a, b []StepStat) []StepStat {
 
 	return merged
 }
+
+// heapItem represents an item in the min-heap for merging step stats
+type heapItem struct {
+	value    StepStat
+	arrayIdx int
+	itemIdx  int
+}
+
+// mergeManySamplesProcessedPerStep merges multiple sorted arrays of StepStat using a min-heap.
+func mergeManySamplesProcessedPerStep(arrays ...[]StepStat) []StepStat {
+	// Filter out empty arrays
+	nonEmptyArrays := make([][]StepStat, 0, len(arrays))
+	totalSize := 0
+	for _, arr := range arrays {
+		if len(arr) > 0 {
+			nonEmptyArrays = append(nonEmptyArrays, arr)
+			totalSize += len(arr)
+		}
+	}
+
+	// Handle edge cases
+	if len(nonEmptyArrays) == 0 {
+		return nil
+	}
+	if len(nonEmptyArrays) == 1 {
+		return nonEmptyArrays[0]
+	}
+	if len(nonEmptyArrays) == 2 {
+		return mergeSamplesProcessedPerStep(nonEmptyArrays[0], nonEmptyArrays[1])
+	}
+
+	// Initialize heap with first element from each array
+	heap := make([]heapItem, len(nonEmptyArrays))
+	for i, arr := range nonEmptyArrays {
+		heap[i] = heapItem{
+			value:    arr[0],
+			arrayIdx: i,
+			itemIdx:  0,
+		}
+	}
+
+	// Build min-heap
+	buildMinHeap(heap)
+
+	merged := make([]StepStat, 0, totalSize)
+
+	for len(heap) > 0 {
+		// Extract minimum element
+		minItem := heap[0]
+
+		// For same timestamp, take the later value (last array wins)
+		if len(merged) > 0 && merged[len(merged)-1].Timestamp == minItem.value.Timestamp {
+			merged[len(merged)-1] = minItem.value
+		} else {
+			merged = append(merged, minItem.value)
+		}
+
+		// Move to next element in the same array
+		nextIdx := minItem.itemIdx + 1
+		if nextIdx < len(nonEmptyArrays[minItem.arrayIdx]) {
+			heap[0] = heapItem{
+				value:    nonEmptyArrays[minItem.arrayIdx][nextIdx],
+				arrayIdx: minItem.arrayIdx,
+				itemIdx:  nextIdx,
+			}
+			heapifyDown(heap, 0)
+		} else {
+			// Remove this array from heap by replacing with last element
+			heap[0] = heap[len(heap)-1]
+			heap = heap[:len(heap)-1]
+			if len(heap) > 0 {
+				heapifyDown(heap, 0)
+			}
+		}
+	}
+
+	return merged
+}
+
+// buildMinHeap builds a min-heap in-place
+func buildMinHeap(heap []heapItem) {
```

Comment:
let's use the `container/heap` package instead of implementing our own heap.

Reply:
Thanks